homeassistant/components/sensor/recorder.py

```diff
@@ -0,0 +1,92 @@
+"""Statistics helper for sensor."""
+from __future__ import annotations
+
+import datetime
+import logging
+import statistics
+
+from homeassistant.components import history
+from homeassistant.const import ATTR_DEVICE_CLASS
+from homeassistant.core import HomeAssistant
+
+from . import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+DEVICE_CLASSES = {"temperature"}
+DEVICE_CLASS_STATISTICS = {"temperature": {"mean", "min", "max"}}
+
+
+def _get_entities(hass: HomeAssistant) -> list[tuple[str, str]]:
+    """Get (entity_id, device_class) of all sensors for which to compile statistics."""
+    all_sensors = hass.states.async_entity_ids(DOMAIN)
+    entity_ids = []
+
+    for entity_id in all_sensors:
+        state = hass.states.get(entity_id)
+        if not state:
+            continue
+        device_class = state.attributes.get(ATTR_DEVICE_CLASS)
+        if not device_class or device_class not in DEVICE_CLASSES:
+            continue
+        entity_ids.append((entity_id, device_class))
+    return entity_ids
+
+
+# Faster than try/except
+# From https://stackoverflow.com/a/23639915
+def _is_number(s: str) -> bool:
+    """Return True if string is a number."""
+    return s.replace(".", "", 1).isdigit()
+
+
+async def async_compile_statistics(
```

Comment:
Can we have a sync version and async version so we can run the sync code in the recorder thread and avoid context switches?

This looks like it will do
`Recorder -> async -> executor -> async -> Recorder`

Reply:
Improved in bdd38f783c0bf03fcb66cc32d7f3c62b62b7ebc2, there's still one context switch though because we need to access the statemachine.